src/statemachine/statemachine_coroutine.js

```diff
@@ -0,0 +1,225 @@
+pc.extend(pc, (function () {
```

Comment:
Further thoughts on this.  It feels like to avoid the singleton issue we should switch my pc.interpolate to be just the inline ones. 

We would then add an app.coroutine call to create a new coroutine and associate it with the manager.  For over time interpolation it gets more tricky: we can provide an interpolate namespace on Coroutine to do coroutine based interpolations.  In this way pc.interpolate needs to know nothing about app and coroutines and becomes a utility for the others - however of course, this pushes interpolate towards core functionality if it comes out of a coroutine - now in core - which I’d guess you might want to avoid.  This is a primary use case though for less skilled developers so I’d guess we want something really obvious.

The only other choice would appear to be passing a coroutine or an app to pc.interpolate.overTime.xxx which feels cumbersome.

So perhaps?

	app.coroutine.startCoroutine(fn, opts)
	app.coroutine.interpolate.easeInOut(start, end, time, fn);


We can obviously keep

	entity.startCoroutine(fn, opts) 

As this already knows it’s app


We could go all Unity and only allow coroutines on Entities but this causes its own set of challenges and a vast amount of confusion on Unity Answers.


> On 22 Jun 2015, at 16:40, Michael Talbot <mike.talbot@3radical.com> wrote:
> 
> Just doing this.  In pc.Coroutine - what would be the right way of getting a coroutine manager for the current app? Pass app as a parameter? Use pc.Application.getApplication()?
>> On 22 Jun 2015, at 13:11, David Evans <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> In src/statemachine/statemachine_coroutine.js <https://github.com/playcanvas/engine/pull/326#discussion_r32927475>:
>> 
>> > @@ -0,0 +1,225 @@
>> > +pc.extend(pc, (function () {
>> In framework_application.js in the constructor. I'd do something like this:
>> 
>> this._statemachineManager = new pc.StateMachineManager();
>> Then in update()
>> 
>> You could just call
>> 
>> this._stateMachineManager.update(dt);
>> —
>> Reply to this email directly or view it on GitHub <https://github.com/playcanvas/engine/pull/326/files#r32927475>.
>> 
> 


-- 
See our 2 minute video on DBS bringing gamification to Singapore here: 
www.3radical.com/resources/videos/#DBSVideo


3radical Limited, 1 Friary, Temple Quay, Bristol BS1 6EA, England.  
T: +44 117 344 5008 ¦ F : +44 117 344 5188

3radical Pte Ltd, 3 Church Street, Level 8 Samsung Hub, Singapore 049483.  
T: +65 6408 3711 ¦ F : +65 6408 0101

www.3radical.com

3radical Limited is a limited company registered in England and Wales 
under number 7872556.  The registered address of 3radical Limited is 1 
Friary, Temple Quay, Bristol BS1 6EA, England.

3radical Pte Ltd is a registered in Singapore under number 201312239K.  The 
registered address of 3radical Pte Ltd is 51 Goldhill Plaza, #07-10/11, 
Singapore 308900.

The information in this e-mail and any attachments may be confidential.  If 
you are not the intended recipient, please notify the sender as soon as 
possible and delete any copies. Unauthorised use, dissemination, 
distribution, publication or copying of this communication is prohibited.

Attachments to this email have been checked with virus detection 
software before transmission, however you should carry out virus checks 
before opening any attachment. 3radical Limited and 3radical Pte Ltd 
accepts no liability for any loss or damage which may be caused by software 
viruses.